File: xianyu_backend/app/schemas/video_schema.py
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from app.schemas.common_schema import PaginationParams


class VideoBase(BaseModel):
    """视频基础模型"""
    title: str = Field(..., description="视频标题", max_length=255)
    type: Optional[str] = Field(None, description="视频类型：电影/电视剧/动漫/短剧等", max_length=100)
    cover_url: Optional[str] = Field(None, description="封面图片URL")
    rating: Optional[float] = Field(None, ge=0, le=10, description="视频评分，0-10分")
    quality_tag: Optional[str] = Field(None, description="视频质量标签：1080P/720P/480P/WEB-DL等", max_length=100)
    description: Optional[str] = Field(None, description="视频简短描述")
    episode_count: Optional[int] = Field(default=0, description="视频总集数")
    actors: Optional[str] = Field(None, description="演员列表，多个演员用逗号分隔")
    year: Optional[int] = Field(None, description="上映年份")
    baidu_url: Optional[str] = Field(None, description="百度网盘资源链接")
    quark_url: Optional[str] = Field(None, description="夸克网盘资源链接")
    other_cloud_url: Optional[str] = Field(None, description="其他云盘资源链接")
    url_status: int = Field(default=1, description="网盘链接状态：0-都失效，1-主流有效，2-其他有效，3-都有效")
    marketing_text: Optional[str] = Field(None, description="营销文本：闲鱼等平台推广文本")
    extra_params: Optional[Dict[str, Any]] = Field(None, description="额外参数，存放对应type的专属属性")


class VideoCreate(VideoBase):
    """创建视频请求模型"""
    pass


class VideoUpdate(BaseModel):
    """更新视频请求模型"""
    title: Optional[str] = Field(None, description="视频标题", max_length=255)
    type: Optional[str] = Field(None, description="视频类型：电影/电视剧/动漫/短剧等", max_length=100)
    cover_url: Optional[str] = Field(None, description="封面图片URL")
    rating: Optional[float] = Field(None, ge=0, le=10, description="视频评分，0-10分")
    quality_tag: Optional[str] = Field(None, description="视频质量标签：1080P/720P/480P/WEB-DL等", max_length=100)
    description: Optional[str] = Field(None, description="视频简短描述")
    episode_count: Optional[int] = Field(None, description="视频总集数")
    actors: Optional[str] = Field(None, description="演员列表，多个演员用逗号分隔")
    year: Optional[int] = Field(None, description="上映年份")
    baidu_url: Optional[str] = Field(None, description="百度网盘资源链接")
    quark_url: Optional[str] = Field(None, description="夸克网盘资源链接")
    other_cloud_url: Optional[str] = Field(None, description="其他云盘资源链接")
    url_status: Optional[int] = Field(None, description="网盘链接状态：0-都失效，1-主流有效，2-其他有效，3-都有效")
    marketing_text: Optional[str] = Field(None, description="营销文本：闲鱼等平台推广文本")
    extra_params: Optional[Dict[str, Any]] = Field(None, description="额外参数，存放对应type的专属属性")


class VideoResponse(VideoBase):
    """视频响应模型"""
    id: int = Field(..., description="视频ID")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: Optional[datetime] = Field(None, description="更新时间")
    deleted_at: Optional[datetime] = Field(None, description="删除时间")

    class Config:
        from_attributes = True


class VideoListResponse(BaseModel):
    """视频列表响应模型"""
    items: List[VideoResponse] = Field(..., description="视频列表")
    total: int = Field(..., description="总数量")
    page: int = Field(..., description="当前页码")
    size: int = Field(..., description="每页数量")


class VideoQueryParams(PaginationParams):
    """视频分页查询参数（仅用于基础分页）"""
    type: Optional[str] = Field(None, description="视频类型过滤")


class VideoConditionParams(PaginationParams):
    """视频条件查询参数（用于高级条件搜索）"""
    title: Optional[str] = Field(None, description="标题关键词搜索")
    type: Optional[str] = Field(None, description="视频类型筛选")
    url_status: Optional[int] = Field(None, description="网盘链接状态筛选")
    actors: Optional[str] = Field(None, description="演员关键词搜索")
    marketing_text: Optional[str] = Field(None, description="营销文本关键词搜索")
    year: Optional[int] = Field(None, description="上映年份筛选")
    rating: Optional[float] = Field(None, ge=0, le=10, description="评分筛选，0-10分")
    quality_tag: Optional[str] = Field(None, description="视频质量标签筛选：1080P/720P/480P/WEB-DL等")

File: xianyu_frontend/src/features/video/schemas/video-schema.ts
export interface Video {
  id: number
  title: string
  type?: string | null
  cover_url?: string | null
  rating?: number | null
  quality_tag?: string | null
  description?: string | null
  episode_count?: number | null
  actors?: string | null
  year?: number | null
  baidu_url?: string | null
  quark_url?: string | null
  other_cloud_url?: string | null
  url_status: number
  marketing_text?: string | null
  extra_params?: Record<string, unknown> | null
  region?: string | null
  finished?: number | null
  sale_status?: number | null
  hot_tags?: string | null
  hot_score?: number | null
  hot_reason?: string | null
  created_at: string
  updated_at?: string | null
  deleted_at?: string | null
}

export interface VideoListResponse {
  items: Video[]
  total: number
  page: number
  size: number
}

export interface VideoQueryParams {
  page?: number
  size?: number
  type?: string
  title?: string
  url_status?: number
  actors?: string
  marketing_text?: string
  year?: number
  rating?: number
  quality_tag?: string
}

export interface VideoCreate {
  title: string
  type?: string
  cover_url?: string
  rating?: number
  quality_tag?: string
  description?: string
  episode_count?: number
  actors?: string
  year?: number
  baidu_url?: string
  quark_url?: string
  other_cloud_url?: string
  url_status?: number
  marketing_text?: string
  extra_params?: Record<string, unknown>
}

export type VideoUpdate = Partial<VideoCreate>

export interface VideoSearchValues {
  title: string
  type: string
  url_status: string
  actors: string
  year: string
  quality_tag: string
}

export interface VideoFormValues {
  title: string
  type: string
  cover_url: string
  rating: string
  quality_tag: string
  description: string
  episode_count: string
  actors: string
  year: string
  baidu_url: string
  quark_url: string
  other_cloud_url: string
  url_status: number
  marketing_text: string
}

export const videoTypeOptions = ['电影', '电视剧', '动漫', '短剧'] as const

export const urlStatusOptions = [
  { value: 0, label: '全部失效' },
  { value: 1, label: '主流有效' },
  { value: 2, label: '其他有效' },
  { value: 3, label: '全部有效' },
] as const

export const emptyVideoSearchValues: VideoSearchValues = {
  title: '',
  type: '',
  url_status: '',
  actors: '',
  year: '',
  quality_tag: '',
}

export function createVideoFormValues(video?: Video | null): VideoFormValues {
  return {
    title: video?.title ?? '',
    type: video?.type ?? '',
    cover_url: video?.cover_url ?? '',
    rating: video?.rating === null || video?.rating === undefined ? '' : String(video.rating),
    quality_tag: video?.quality_tag ?? '',
    description: video?.description ?? '',
    episode_count:
      video?.episode_count === null || video?.episode_count === undefined ? '' : String(video.episode_count),
    actors: video?.actors ?? '',
    year: video?.year === null || video?.year === undefined ? '' : String(video.year),
    baidu_url: video?.baidu_url ?? '',
    quark_url: video?.quark_url ?? '',
    other_cloud_url: video?.other_cloud_url ?? '',
    url_status: video?.url_status ?? 1,
    marketing_text: video?.marketing_text ?? '',
  }
}

function cleanText(value: string) {
  const trimmedValue = value.trim()
  return trimmedValue.length > 0 ? trimmedValue : undefined
}

function cleanNumber(value: string) {
  const trimmedValue = value.trim()

  if (!trimmedValue) {
    return undefined
  }

  const numberValue = Number(trimmedValue)
  return Number.isFinite(numberValue) ? numberValue : undefined
}

function cleanInteger(value: string) {
  const numberValue = cleanNumber(value)
  return numberValue === undefined ? undefined : Math.trunc(numberValue)
}

export function videoFormValuesToPayload(values: VideoFormValues): VideoCreate {
  const payload: VideoCreate = {
    title: values.title.trim(),
    url_status: values.url_status,
  }
  const type = cleanText(values.type)
  const coverUrl = cleanText(values.cover_url)
  const rating = cleanNumber(values.rating)
  const qualityTag = cleanText(values.quality_tag)
  const description = cleanText(values.description)
  const episodeCount = cleanInteger(values.episode_count)
  const actors = cleanText(values.actors)
  const year = cleanInteger(values.year)
  const baiduUrl = cleanText(values.baidu_url)
  const quarkUrl = cleanText(values.quark_url)
  const otherCloudUrl = cleanText(values.other_cloud_url)
  const marketingText = cleanText(values.marketing_text)

  if (type) payload.type = type
  if (coverUrl) payload.cover_url = coverUrl
  if (rating !== undefined) payload.rating = rating
  if (qualityTag) payload.quality_tag = qualityTag
  if (description) payload.description = description
  if (episodeCount !== undefined) payload.episode_count = episodeCount
  if (actors) payload.actors = actors
  if (year !== undefined) payload.year = year
  if (baiduUrl) payload.baidu_url = baiduUrl
  if (quarkUrl) payload.quark_url = quarkUrl
  if (otherCloudUrl) payload.other_cloud_url = otherCloudUrl
  if (marketingText) payload.marketing_text = marketingText

  return payload
}

export function videoSearchValuesToParams(values: VideoSearchValues, pageSize: number): VideoQueryParams {
  const year = cleanInteger(values.year)

  return {
    page: 1,
    size: pageSize,
    title: cleanText(values.title),
    type: cleanText(values.type),
    url_status: values.url_status ? Number(values.url_status) : undefined,
    actors: cleanText(values.actors),
    year,
    quality_tag: cleanText(values.quality_tag),
  }
}
